Create CrowdSignal survey views and navigation

ThomazFB opened the pull request:
Summary
==========
Fixes issue #2644, this PR delivers the following features:
- Trigger Crowdsignal Feedback Survey when `Could be better` button is clicked on the `FeedbackRequestCard`
- Crowdsignal Feedback Survey view using a WebView
- WebView lifecycle handling with loading until survey view is completely loaded and survey completion interception
- Feedback Completed view with Dark Mode support
- Feedback fragments navigation, back stack and exit handled through Navigation Controller
- Contact Us text button integrated with `Help & Support` view

Changelog
==========
| Description | Commit |
| ------------- | ------------- |
| Initiate Survey fragments implementation | d0716b4d3d8290da28bbed92fbf200a17fb513c5 |
| Integrate Survey fragments with the Navigation Controller declaration structure |  9f3272e28df6da4572918cacf342d9570b867f64 |
| Create interception of the WebView lifecycle to make possible the loading dialog presentation and removal | e08773639d83e2425390b3b889ec247b3909386f |
| Make JavaScript execution available inside the WebView in order to better fit CrowdSignal web navigation experience | e55f021173383a6f2fb4b90b1a2cb0a78aad8ab5 |
| Create rule to intercept URL query parameter interaction through request loading override to verify if CrowdSignal survey is complete | 440f3c8fb9fb4d07829646ddf236f581cb4570b7 |
| `FeedbackCompletedFragment` created | db0dbad1b423964cde5df9d49acc2c5ee32b673d |
| `FeedbackCompletedFragment` UI structure defined | 530b3c6ac825e3766c1ab9531fdf11df5b06f786 |
| Change Survey URL declaration to the `AppUrls` object | 002f7f15062b54fa471d9f3e6d7f2f9dc103396b |
| Make the Crowdsignal Feedback Survey navigation start available from the `DashboardFragment` view | 2398d88a241251b4a1ea4933a89465445185291f |

Screenshots
==========
Light Mode
------------
| Click on `Could be better` button | Wait for the survey loading | Interact with the survey until it's finished | Verify if the expected view is presented and is fully functional | 
| ------------- | ------------- | ------------- | ------------- |
| ![Screenshot_1597814080](https://user-images.githubusercontent.com/5920403/90594826-c2b80900-e1c1-11ea-8368-af5b36c23dff.png) | ![Screenshot_1597812682](https://user-images.githubusercontent.com/5920403/90594829-c5b2f980-e1c1-11ea-9fb4-c8e04118efe3.png) | ![Screenshot_1597891397](https://user-images.githubusercontent.com/5920403/90712144-bcd12f00-e278-11ea-9b6c-b1b014139d00.png) | ![Screenshot_1597812776](https://user-images.githubusercontent.com/5920403/90594835-c9df1700-e1c1-11ea-9b16-63bdb680ed8c.png) |




Dark Mode
------------
| Click on `Could be better` button | Wait for the survey loading | Interact with the survey until it's finished | Verify if the expected view is presented and is fully functional | 
| ------------- | ------------- | ------------- | ------------- |
| ![Screenshot_1597891515](https://user-images.githubusercontent.com/5920403/90711653-a70f3a00-e277-11ea-96d8-f094ca3a722d.png) | ![Screenshot_1597891046](https://user-images.githubusercontent.com/5920403/90711657-a8d8fd80-e277-11ea-97d5-79f0a99c178d.png) | ![Screenshot_1597891102](https://user-images.githubusercontent.com/5920403/90711659-aa0a2a80-e277-11ea-9d2b-a462a141dcd6.png) | ![Screenshot_1597891109](https://user-images.githubusercontent.com/5920403/90711662-ab3b5780-e277-11ea-84c9-83b23f977fc0.png) |

Tablet
------------
| Click on `Could be better` button | Wait for the survey loading | Interact with the survey until it's finished | Verify if the expected view is presented and is fully functional | 
| ------------- | ------------- | ------------- | ------------- |
| ![Screenshot_1597891681](https://user-images.githubusercontent.com/5920403/90711694-cb6b1680-e277-11ea-92e0-5c83e858fcc2.png) | ![Screenshot_1597891688](https://user-images.githubusercontent.com/5920403/90711703-cc9c4380-e277-11ea-8911-e44f05a40bd9.png) | ![Screenshot_1597891695](https://user-images.githubusercontent.com/5920403/90711705-ce660700-e277-11ea-991c-b0ca7388a3c0.png) | ![Screenshot_1597891708](https://user-images.githubusercontent.com/5920403/90711709-cf973400-e277-11ea-8dab-671f24a8aa11.png) |

How to Test
------------

1. Access the `My Store` section from the Woo app
2. Inside the `FeedbackRequestCard`, click in the `Could be better` button
3. Verify that the survey view is called, the loading dialog is visible until the survey is completely loaded
4. Interact with the survey until the conclusion, after finishing it, verify that the custom native Feedback Completed View is called instead of letting the WebView show the default web one.
5. Once inside the Feedback Completed View, verify that the each button is working as expected and described


Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Create basic Fragment structure

ThomazFB committed:
Merge branch 'develop' into issue/2644_add_crowdsignal_survey

ThomazFB committed:
Merge branch 'develop' into issue/2644_add_crowdsignal_survey

ThomazFB committed:
Merge branch 'develop' into issue/2644_add_crowdsignal_survey

ThomazFB committed:
create navigation link for the FeedbackSurveyFragment

ThomazFB committed:
Add Progress Dialog and WebViewClient into FeedbackSurveyFragment

ThomazFB committed:
TabLayout removed when calling FeedbackSurveyFragment

ThomazFB committed:
Enable JavaScript on WebView

ThomazFB committed:
Intercept survey completion

ThomazFB committed:
Create FeedbackCompletedFragment

ThomazFB committed:
Refactor FeedbackSurvey navigation call

ThomazFB committed:
Create navigation link between FeedbackSurveyFragment and FeedbackCompletedFragment

ThomazFB committed:
Add string resources to loading progress dialog

ThomazFB committed:
FeedbackCompletedFragment basic UI frame created

ThomazFB committed:
Attribution adjusted

ThomazFB committed:
Improve FeedbackCompletedFragment dimension definitions

ThomazFB committed:
Define popup behavior when leaving the FeedbackCompletedFragment

ThomazFB committed:
Define CrowdSignal URL inside AppUrls

ThomazFB committed:
Make FeedbackCompletedFragment content text clickable

ThomazFB committed:
Handle options menu lifecycle

ThomazFB committed:
Improve fragment_feedback_completed dimensions definitions

ThomazFB committed:
Merge branch 'develop' into issue/2644_add_crowdsignal_survey

ThomazFB committed:
Merge branch 'develop' into issue/2644_add_crowdsignal_survey

ThomazFB committed:
Add feedback survey call to the DashboardFragment

ThomazFB committed:
Add RELEASE-NOTES description for CrowdSignal survey feature

Garance91540 reviewed (changes requested):
Hi Thomaz, looking good! 

I have a few comments/questions:
- Can you please share with me dark mode screenshots and one iPad screenshot too?
- Which survey are you using exactly? I think we should use the same on Android and iOS so we only have on place to go to see all the data. Here's the Crowdsignal survey @shiki worked on https://automattic.survey.fm/woo-app-general-feedback-test-survey, can we use it for Android too?

ThomazFB commented:
Hey @Garance91540! Thanks for your inputs!
>Can you please share with me dark mode screenshots and one iPad screenshot too?

Sure! I've added into the PR description the navigation sequence running in Dark Mode and on an Android Tablet 😄 

>Which survey are you using exactly? I think we should use the same on Android and iOS so we only have on place to go to see all the data. Here's the Crowdsignal survey @shiki worked on https://automattic.survey.fm/woo-app-general-feedback-test-survey, can we use it for Android too?

I've intentionally left the PR with the test survey URL just for review purposes, but to make easier the design review I've just added the production URL with the CSS changes made by @shiki. Thanks for pointing that out!

Let me know if these changes works for you!

ThomazFB committed:
Update CrowdSignal survey to the production one

Garance91540 reviewed (dismissed):
Looking awesome, thank you @ThomazFB !

shiki commented:
Regarding the style, looks like the radio buttons are squares instead of circles. 😄 I'll fix those next sprint. (https://github.com/woocommerce/woocommerce-ios/issues/2546)

AmandaRiu reviewed (changes requested):
@ThomazFB Nice work on this 🥳 Here's some feedback from what I found during testing:

- I noticed that if the device is rotated the survey loses it's values. Is there a way we can prevent that from happening?
- The "Feedback sent" screen does not scroll so you can't see the rest of the view in landscape mode or click the "Back to Store" button:

<img src="https://user-images.githubusercontent.com/5810477/90783352-f4b69180-e2cd-11ea-91ef-e456c81cf537.png" width="600"/>

@shiki I noticed the survey errors look like this on Android, is this expected?

<img src="https://user-images.githubusercontent.com/5810477/90786104-2da43580-e2d1-11ea-8664-78ea1a18f642.png" width="330"/>

shiki commented:
@AmandaRiu Yes, it's a problem in Crowdsignal's CSS editor. I reported it to them (p91TBi-2RL-p2#comment-1530). Thank you for confirming. 🙂

ThomazFB committed:
Refactor FeedbackSurveyFragment to support webView state restoration

ThomazFB committed:
Create extension function to properly setup toolbar close button

ThomazFB committed:
Make fragment_feedback_completed scrollable

ThomazFB committed:
Refactor null checks

ThomazFB commented:
Hey @AmandaRiu! Thank you so much for your review, I've submitted some change and I think the code is ready for a second round!

>I noticed that if the device is rotated the survey loses it's values. Is there a way we can prevent that from happening?

Thank you for pointing that out! I did some investigations regarding WebView state management and it doesn't look very reliable, nevertheless, I did some additional changes and restored my previous attempts to restore the WebView previous state and it looks to be working now. Let me know how that goes for you. Here's the changes: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/1034133023f225423111c47cd92ea72f01c965a0

>The "Feedback sent" screen does not scroll so you can't see the rest of the view in landscape mode or click the "Back to Store" button

Sorry about that! Here's the change allowing scrolls on the UI: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/7803c858618345e84f8adc5d2276f7189875fcc9

ThomazFB committed:
Remove unused import from FeedbackCompletedFragment

ThomazFB committed:
Rollback previous fragment closing action to default one

AmandaRiu commented on WooCommerce/src/main/res/layout/fragment_feedback_completed.xml:
```diff
@@ -0,0 +1,79 @@
+        android:layout_height="wrap_content"
+        android:gravity="center_horizontal|top"
+        android:orientation="vertical"
+        android:padding="32dp">
```
np: hard code values. Should use one of our established values in `dimens.xml` - same for the rest of this file.

ThomazFB replied:
Sure! Thanks for that! It's done right here: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/c1ad309bd828fa668d181fb13f41ce708841a6b1

AmandaRiu commented on WooCommerce/src/main/res/layout/fragment_feedback_completed.xml:
```diff
@@ -0,0 +1,79 @@
+        android:orientation="vertical"
+        android:padding="32dp">
+
+        <TextView
```
should use a `MaterialTextView`

ThomazFB replied:
Oops, forgot to adjust that one, it's fixed right here: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/c1ad309bd828fa668d181fb13f41ce708841a6b1

AmandaRiu commented on WooCommerce/src/main/res/layout/fragment_feedback_completed.xml:
```diff
@@ -0,0 +1,79 @@
+        <TextView
+            android:id="@+id/completion_title"
+            style="@style/Woo.TextView.Subtitle1"
+            android:layout_width="241dp"
+            android:layout_height="48dp"
```
Why hard-coded widths and heights?

ThomazFB replied:
My mistake here, sorry about that, I've fixed this for all the missing extraction dimensions right here: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/c1ad309bd828fa668d181fb13f41ce708841a6b1

AmandaRiu reviewed (changes requested):
@ThomazFB The scrolling thing is fixed 👍 thanks for taking care of that. I'm still seeing some issues when rotating the survey. The email disappears:

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/5810477/90844378-c10d5300-e331-11ea-83f5-a80c86d2dc72.gif)

Other than that I left a couple smaller comments in code. It's really close to merging! 👍

ThomazFB committed:
Merge branch 'develop' into issue/2644_add_crowdsignal_survey

ThomazFB committed:
Remove hardcoded dimens to pre-defined ones

ThomazFB commented:
Hey @AmandaRiu, thank you again for your review. I've pushed all the missing dimension extractions on the `xml` files. But there's one thing I'm worried about the `WebView` state handling. I did a lot of attempts trying to find a reliable way to restore the state on the `email` field, but it looks like this field uses JS to control their state because of pattern matching to verify if it's a valid e-mail address, and since the Android WebView instance state storing bundle doesn't cover JS variable states and logic control, we see this kind of behavior happening.

Because of that same characteristic of the survey webpage, if we access the [Survey link](https://automattic.survey.fm/woo-app-general-feedback-test-survey) from a desktop browser like Chrome, we can see that even in that case the browser has a hard time to keep the email state in scenarios like pressing back and forward buttons on the browser tab.

I think we'll need to assume some unwanted compromises in order to achieve a good state handling, like blocking the Activity from recreating itself when we do a configuration change like rotation, but I think that's a way too harsh measure since it will affect a lot of other Fragments and Views, and overall it's a not good practice to be adopted. Not sure if we have many options to solve this problem. WDYT?

I'm curious if the iOS part managed to solve this problem in some way. CC: @shiki @Ecarrion

Ecarrion commented:
@ThomazFB 
> I'm curious if the iOS part managed to solve this problem in some way.

I just tested it and it doesn't seem to be a problem iOS side, I assume it's because iOS doesn't recreate view upon rotation, just re-layout them.

AmandaRiu commented:
> I think we'll need to assume some unwanted compromises in order to achieve a good state handling, like blocking the Activity from recreating itself when we do a configuration change like rotation, but I think that's a way too harsh measure since it will affect a lot of other Fragments and Views, and overall it's a not good practice to be adopted. Not sure if we have many options to solve this problem. WDYT?

I agree that preventing a configuration change on rotation would be too harsh for something like this. I think we should leave it for now. The good news is that the survey will not be submitted without the email form filled so the user will be asked to re-enter it if a configuration change happens. And at least the rest of the survey is 👍

ThomazFB committed:
Adjust fragment_feedback_completed TextView width to comply with app standard

ThomazFB committed:
Merge branch 'develop' into issue/2644_add_crowdsignal_survey

AmandaRiu reviewed (approved):
LGTM! :shipit:
